### PR TITLE
sql: sanitize the "no current db" story

### LINKED ISSUE #23145: sql: reduce (eliminate?) support for sessions without a current database set
Discussed with @nvanbenschoten and @jordanlewis while working on #21456 and #23045.

pg does not support client sessions without a current database set. The virtual schemas (especially pg_catalog) are not well-defined if there is no current database. 

Also in their current implementation they can generate a large amount data if queried by a client without a current database set, putting pressure on the SQL allocation pool.

The discussion on #21456 suggests evolving CockroachDB so that client sessions more often find themselves with a current database set by default. There's a spectrum of approaches:

- have the current database set to `system` if not specified in the conn string / pgwire options
- have the current database set to some arbitrary constant (e.g. `default`) and encourage the user via `cockroach sql` to create the database if it doesn't exist yet
  - might be a UX issue since only admin users can create databases
- have the current database set to the username (like in pg), encourage the user to create
  - same UX problem as above
- make every new cluster contain a "default" database, set that as default current database for new sessions that don't specify otherwise
- same as above, but create the "default" db on-demand automatically

cc @bdarnell

### LINKED ISSUE #23893: sql: errors in client due to use of pg_catalog with non-existent db as current
It seems there is a regression with the latest v2 alpha (20180312) as  I am now seeing the following error when Ecto is trying to create the DB from scratch:

```
** (Postgrex.Error) ERROR 3D000 (invalid_catalog_name): database "postgres" does not exist
    (db_connection) lib/db_connection/connection.ex:148: DBConnection.Connection.connect/2
    (connection) lib/connection.ex:622: Connection.enter_connect/5
    (stdlib) proc_lib.erl:247: :proc_lib.init_p_do_apply/3
```

### PR DESCRIPTION
**Pull request by knz**
Fixes #23893.
Fixes #23145.
Updates #24598.
Informs #24056.
Informs #23958.

Prior to this patch, CockroachdB made it excessively simple for client
connections to not have any current database set. Also by design
CockroachDB allows a connection to be opened with a non-existent
database set as current.

This in turn causes various problems related to name resolution and
the `pg_catalog` schema.

To understand these problems we can consider separately two groups of
users, which have separate use cases:

- newcomers to CockroachDB that don't have prior experience with SQL
  or PostgreSQL. These will not know about "databases" in any sense,
  and will use CockroachDB with default options. This includes using
  `cockroach sql` with no current database selected. These users have
  to be taught upfront that they need to "create a database" and "set
  the current database", which they may or may not (more likely not)
  be interested in.

  When these users then transition into writing apps, they will
  copy-paste the conn URL printed by `cockroach start` (which sets no
  current db) into some ORM or framework, and then will regretfully
  observe that their app fail in mysterious ways because the current
  db is not set (also see the next point).

- users of existing applications or SQL drivers that have been
  developed to target PostgreSQL. These apps commonly contain code
  that:

  - connect to a database named "`postgres`" by default, and make this
    default hard(er) to customize. For example Ecto (Elixir ORM)
    doesn't make this configurable.

    (From the PostgreSQL docs: "After initialization, a database
    cluster will contain a database named `postgres`, which is meant as
    a default database for use by utilities, users and third party
    applications. The database server itself does not require the
    postgres database to exist, but many external utility programs
    assume it exists.")

  - (regardless of which db is selected as current), will issue some
    introspection queries using `pg_catalog` before any additional
    initialization is made by the higher-level app code, in particular
    before the app can issue `CREATE DATABASE`. Currently `pg_catalog`
    queries (and several other things) fail if the current database is
    unset or set to a non-existing database.

To address this relatively large class of problems, this patch
modifies CockroachDB as follows:

- all clusters (existing or new) now get two empty databases on
  startup (created by a migration) called `defaultdb` and `postgres`.

- when a client doesn't specify a current db in their conn string
  (i.e. in the pgwire options), the `defaultdb` database is picked up
  instead.

This resolves all the known problems around the situations identified
above.

The two new databases behave like any regular database and are
provided for convenience to newcomers and
compatibility. (Administrators are free to drop them afterwards to
restore the previous situation, if that is actively desired.)

In addition, to make it slightly harder for newcomers to shoot
themselves in the foot, the `database` session variable cannot be set
to the empty string any more if `sql_safe_updates` is otherwise set.

Three alternatives to this approach were considered, discussed with
@jordanlewis and @nvanbenschoten in their quality of pg compatibility
experts, and with @bdarnell in his quality of carer for newcomers, and
ultimately rejected:

A) generate the current db upon connection if it does not exist, make
   connections using the empty string as curdb use `system` instead.

   Rejected because we don't like to auto-create upon connection. Too
   easy to flood the system with bogus database names by typos in the
   connection string.

B) Pre-populate clusters with a special anonymous database (db name =
   empty string).

   Rejected because there are too many areas in CockroachDB with a
   risk to fail silently or with weird error messages as a
   result. Also does not solve the problem of clients that want a db
   called `postgres`.

C) Auto-create a database named after the user upon user creation
   (with an optional flag to skip db creation).

   Rejected because it does not solve the problem of clients that want
   a db called `postgres`. @bdarnell: creating a database per user also
   implicitly encourages bad habits (like sharing certs for the user
   whose name corresponds to the database instead of creating multiple
   users).

D) Implement some magic in the handling of `pg_catalog`, name
   resolution and all other pieces that currently don't like non-existent
   `database` values to create the appearance of something that works.

   Rejected because too complex to implement successfully (there are
   many moving parts that need to be touched to make this work) and
   the resulting magic would be hard to understand by CockroachDB
   contributors and hard to maintain over time.

E) Auto-create only one database called `defaultdb`.

   Rejected because it does not solve the problem of clients that want
   a db called `postgres`.

F) Auto-create only one database called `postgres` and use that as
   default when no current db is specified.

   Rejected because not good for branding.

G) Name the `defaultdb` either `def`, `default` or `default_database`.

   The word "`default`" was rejected because it is a SQL keyword
   and would cause queries using it to fail in obscure ways.

   The word "`def`" (similar to what MySQL uses in
   `information_schema`) was rejected because it refers too strongly
   to other uses of this word in programming languages ("define").

   "`default_database`" was rejected because too verbose.

Release note (sql change): new and existing clusters will now contain
two empty databases called `defaultdb` and `postgres` by default. The
database `defaultdb` is automatically used for clients that connect
without a current database set (e.g. without a database component in
the connection URL). The database `postgres` is provided for
compatibility with PostgreSQL client frameworks that require it to
exist when the database server has ben freshly installed. Both new
databases behave like any other regular database.

Release note (general change): existing clusters upgraded to this
version of CockroachDB will automatically see two new empty databases
named `defaultdb` and `postgres`. These were added for compatibility
with PostgreSQL and to ease adoption of CockroachDB. They can be
manually deleted by an administrator if the client applications are
determined not to require them.

### CODE REVIEW

**Comment by knz** [comment]
@benesch can we sit together to fix the test failures caused by the appearance of 2 non-system descriptors upon startup of test clusters. There are failures because we now have 2 more ranges than expected, and I don't know where to look.

File: pkg/sqlmigrations/migrations.go [hunk 1]
```diff
@@ -186,6 +187,11 @@ var backwardCompatibleMigrations = []migrationDescriptor{
 		name:   "ensure admin role privileges in all descriptors",
 		workFn: ensureMaxPrivileges,
 	},
+	{
+		// Introduced in v2.0.
+		name:   "create default databases",
+		workFn: createDefaultDbs,
```
Comment (benesch) [inline pkg/sqlmigrations/migrations.go]:
```go

```

**Comment by knz** [comment]
@benesch thank you! I needed to tweak it a little bit but it seems to work :)

**Comment by knz** [comment]
@bdarnell are we going to queue this for 2.0.x? I'll need to know to make the comment on the migration accurrate.

**Comment by bdarnell** [comment]
No, I think this is too big of a behavior change for 2.0.x.

**Comment by knz** [comment]
All right thanks.
The patch is now ready for review.

**Comment by knz** [comment]
rebased to solve conflicts

**Comment by knz** [comment]
Friendly invitation to start the review on this PR and/or delegate the review to other suitable reviewers.

(I know I need to rebase the PR, but it passed all tests before the merge conflict.  I am confident that the rebase will not change the gist of the change.)

**Comment by benesch** [comment]
The expected initial range stuff LGTM (obviously), but the rest of this PR is out of my depth.

**Comment by BramGruneir** [comment]
:lgtm: With a number of small nits and it needs a few more tests.

---

Reviewed 17 of 24 files at r1, 29 of 29 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[pkg/sql/drop_test.go, line 502 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/24735#-LC9PDTrBli99mrO6lqp:-LC9PDTrBli99mrO6lqq:br2a300) ([raw file](https://github.com/cockroachdb/cockroach/blob/1fe06d67c7e5ad91bc471e1f4ecf847af2b90d5f/pkg/sql/drop_test.go#L502)):*
> ```Go
> 
> 	tableDesc := sqlbase.GetTableDescriptor(kvDB, "t", "kv")
> 	nameKey := sqlbase.MakeNameMetadataKey(keys.MaxReservedDescID+3, "kv")
> ```

Not a fan of magic numbers.  Can you add a comment or make it a named const?

---

*[pkg/sql/schema_changer_test.go, line 82 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/24735#-LC9PQuW4RYQlWqDR3wp:-LC9PQuW4RYQlWqDR3wq:brwzlix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1fe06d67c7e5ad91bc471e1f4ecf847af2b90d5f/pkg/sql/schema_changer_test.go#L82)):*
> ```Go
> 
> 	var lease sqlbase.TableDescriptor_SchemaChangeLease
> 	var id = sqlbase.ID(keys.MaxReservedDescID + 4)
> ```

Please add a named const or at least a comment.  Why was this 2 before? 

---

*[pkg/sql/zone_config_test.go, line 95 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/24735#-LC9QkbU6ps73OVY2AVK:-LC9QkbU6ps73OVY2AVL:b1izy2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/1fe06d67c7e5ad91bc471e1f4ecf847af2b90d5f/pkg/sql/zone_config_test.go#L95)):*
> ```Go
> 	s := srv.(*server.TestServer)
> 
> 	expectedCounter := uint32(keys.MaxReservedDescID + 2)
> ```

Again, comment or const.  Maybe a shared const since this seems very common.

---

*[pkg/sql/zone_test.go, line 72 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/24735#-LC9R9_Q5-kx0wgQUwcL:-LC9R9_Q5-kx0wgQUwcM:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/1fe06d67c7e5ad91bc471e1f4ecf847af2b90d5f/pkg/sql/zone_test.go#L72)):*
> ```Go
> 	}
> 	dbRow := sqlutils.ZoneRow{
> 		ID:           keys.MaxReservedDescID + 3,
> ```

ditto

---

*[pkg/sql/logictest/testdata/logic_test/drop_database, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/24735#-LC9RgpUBJ2LCfcBxPEH:-LC9RgpUBJ2LCfcBxPEI:b-lq0cr1) ([raw file](https://github.com/cockroachdb/cockroach/blob/1fe06d67c7e5ad91bc471e1f4ecf847af2b90d5f/pkg/sql/logictest/testdata/logic_test/drop_database#L1)):*
> ```
> # LogicTest: default
> ```

Please add a test that drops both of these new dbs.
And maybe even recreate the,.

---

*[pkg/sql/logictest/testdata/logic_test/grant_table, line 69 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/24735#-LC9SEVM8BVbNN_m-hX3:-LC9SEVM8BVbNN_m-hX4:b-9ibvi2) ([raw file](https://github.com/cockroachdb/cockroach/blob/1fe06d67c7e5ad91bc471e1f4ecf847af2b90d5f/pkg/sql/logictest/testdata/logic_test/grant_table#L69)):*
> ```
> a         pg_catalog          NULL              readwrite  ALL
> a         pg_catalog          NULL              root       ALL
> a         public              NULL              admin      ALL
> ```

Not sure how to review this.  Perhaps we should make this programmatic instead of a list?  But that can be for another PR. 

---

*[pkg/sql/logictest/testdata/logic_test/rename_database, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/24735#-LC9SjyO9WvUyUfcqAkh:-LC9SjyO9WvUyUfcqAki:b-gl7pr9) ([raw file](https://github.com/cockroachdb/cockroach/blob/1fe06d67c7e5ad91bc471e1f4ecf847af2b90d5f/pkg/sql/logictest/testdata/logic_test/rename_database#L1)):*
> ```
> # LogicTest: default parallel-stmts distsql
> ```

Please add a test that renames these new dbs.

---

*[pkg/sql/pgwire/pgwire_test.go, line 704 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/24735#-LC9T8LnD57gHaXGgq7O:-LC9T8LnD57gHaXGgq7P:b-rci90x) ([raw file](https://github.com/cockroachdb/cockroach/blob/1fe06d67c7e5ad91bc471e1f4ecf847af2b90d5f/pkg/sql/pgwire/pgwire_test.go#L704)):*
> ```Go
> 		},
> 		"SET CLUSTER SETTING cluster.organization = $1": {
> 			baseTest.SetArgs("hello world"),
> ```

Would it make sense to add some tests that execute the exact behaviour described in the commit message?

---

*[pkg/sql/sessiondata/search_path.go, line 22 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/24735#-LC9TYeQ7uql6moUHgJw:-LC9TYeQ7uql6moUHgJx:b3e0vms) ([raw file](https://github.com/cockroachdb/cockroach/blob/1fe06d67c7e5ad91bc471e1f4ecf847af2b90d5f/pkg/sql/sessiondata/search_path.go#L22)):*
> ```Go
> 
> // PgDatabaseName is the name of the default postgres system database.
> const PgDatabaseName = "postgres"
> ```

If this DB are renamed, should the serachpath be updated too?

---

*[pkg/sql/sessiondata/search_path.go, line 26 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/24735#-LC9TaoM0J5BFKRRA6so:-LC9TaoM0J5BFKRRA6sp:b-mqvx0f) ([raw file](https://github.com/cockroachdb/cockroach/blob/1fe06d67c7e5ad91bc471e1f4ecf847af2b90d5f/pkg/sql/sessiondata/search_path.go#L26)):*
> ```Go
> // DefaultDatabaseName is the name ofthe default CockroachDB database used
> // for connections without a current db set.
> const DefaultDatabaseName = "default"
> ```

Should this be added to the searchpath?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24735#-:-LC9UH4SER1mfSveCLPN:b-6ttvil)*

**Comment by knz** [comment]
@BramGruneir thanks for pushing me to test the "default" database more thoroughly. It turns out that DEFAULT being a keyword in SQL, all kinds of things break in subtle and non-subtle ways when a query attempts to use this database name unquoted. I think this would cause too much confusion for users (And too much overhead in docs) so I am thinking of a new name for that database. Perhaps "def" as is done in MySQL.

**Comment by knz** [comment]
Review status: 14 of 45 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/sql/pgwire/pgwire_test.go, line 704 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/24735#-LC9T8LnD57gHaXGgq7O:-LD1hzwQAoEsKZlRdZTL:b77vi7k) ([raw file](https://github.com/cockroachdb/cockroach/blob/1fe06d67c7e5ad91bc471e1f4ecf847af2b90d5f/pkg/sql/pgwire/pgwire_test.go#L704)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Would it make sense to add some tests that execute the exact behaviour described in the commit message?
</blockquote></details>

What behavior are you referring to?

---

*[pkg/sql/sessiondata/search_path.go, line 22 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/24735#-LC9TYeQ7uql6moUHgJw:-LD1i30a61AUwOVL57Za:b8jtt96) ([raw file](https://github.com/cockroachdb/cockroach/blob/1fe06d67c7e5ad91bc471e1f4ecf847af2b90d5f/pkg/sql/sessiondata/search_path.go#L22)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

If this DB are renamed, should the serachpath be updated too?
</blockquote></details>

since 2.0 search path is about schemas, not databases.

---

*[pkg/sql/sessiondata/search_path.go, line 26 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/24735#-LC9TaoM0J5BFKRRA6so:-LD1i6LD0qf8vlnq8MMD:b2ss7xm) ([raw file](https://github.com/cockroachdb/cockroach/blob/1fe06d67c7e5ad91bc471e1f4ecf847af2b90d5f/pkg/sql/sessiondata/search_path.go#L26)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Should this be added to the searchpath?
</blockquote></details>

see comment above

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24735)*

**Comment by knz** [comment]
Review status: 9 of 46 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/sql/logictest/testdata/logic_test/drop_database, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/24735#-LC9RgpUBJ2LCfcBxPEH:-LD1pYIBCSo2ON6I6jX1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1fe06d67c7e5ad91bc471e1f4ecf847af2b90d5f/pkg/sql/logictest/testdata/logic_test/drop_database#L1)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Please add a test that drops both of these new dbs.
And maybe even recreate the,.
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/grant_table, line 69 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/24735#-LC9SEVM8BVbNN_m-hX3:-LD1p_vED1GFiecnINq-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1fe06d67c7e5ad91bc471e1f4ecf847af2b90d5f/pkg/sql/logictest/testdata/logic_test/grant_table#L69)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Not sure how to review this.  Perhaps we should make this programmatic instead of a list?  But that can be for another PR. 
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/rename_database, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/24735#-LC9SjyO9WvUyUfcqAkh:-LD1pb_9DSKi-0QGVyuN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1fe06d67c7e5ad91bc471e1f4ecf847af2b90d5f/pkg/sql/logictest/testdata/logic_test/rename_database#L1)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Please add a test that renames these new dbs.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24735)*

**Comment by BramGruneir** [comment]
Reviewed 37 of 38 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/sql/vars.go, line 142 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/24735#-LD2JZrK8tIgvyNG-2em:-LD2JZrK8tIgvyNG-2en:bgy7tjc) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4723a9b3db45c7513a0c87f60f6222317470327/pkg/sql/vars.go#L142)):*
> ```Go
> 
> 			if len(dbName) == 0 && evalCtx.SessionData.SafeUpdates {
> 				return pgerror.NewDangerousStatementErrorf("SET database to empty string")
> ```

Is there a test for this?

---

*[pkg/sql/pgwire/pgwire_test.go, line 704 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/24735#-LC9T8LnD57gHaXGgq7O:-LD2ImH82sn0plsmiuzb:bgok3bm) ([raw file](https://github.com/cockroachdb/cockroach/blob/1fe06d67c7e5ad91bc471e1f4ecf847af2b90d5f/pkg/sql/pgwire/pgwire_test.go#L704)):*
> when a client doesn't specify a current db in their conn string
> (i.e. in the pgwire options), the defaultdb database is picked up
> instead.

Not sure if it's easy to test.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24735)*

**Comment by knz** [comment]
Review status: 38 of 61 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/sql/drop_test.go, line 502 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/24735#-LC9PDTrBli99mrO6lqp:-LD2PzsD8MFITlnH3zY0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1fe06d67c7e5ad91bc471e1f4ecf847af2b90d5f/pkg/sql/drop_test.go#L502)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Not a fan of magic numbers.  Can you add a comment or make it a named const?
</blockquote></details>

Done.

---

*[pkg/sql/schema_changer_test.go, line 82 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/24735#-LC9PQuW4RYQlWqDR3wp:-LD2Q0EY0wAwBDev36aG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1fe06d67c7e5ad91bc471e1f4ecf847af2b90d5f/pkg/sql/schema_changer_test.go#L82)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Please add a named const or at least a comment.  Why was this 2 before? 
</blockquote></details>

Done.

---

*[pkg/sql/zone_config_test.go, line 95 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/24735#-LC9QkbU6ps73OVY2AVK:-LD2Q3RZ1XbmJXpmuy_i:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1fe06d67c7e5ad91bc471e1f4ecf847af2b90d5f/pkg/sql/zone_config_test.go#L95)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Again, comment or const.  Maybe a shared const since this seems very common.
</blockquote></details>

Done.

---

*[pkg/sql/zone_test.go, line 72 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/24735#-LC9R9_Q5-kx0wgQUwcL:-LD2Q4I25MFu06OoBl28:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1fe06d67c7e5ad91bc471e1f4ecf847af2b90d5f/pkg/sql/zone_test.go#L72)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---

*[pkg/sql/pgwire/pgwire_test.go, line 704 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/24735#-LC9T8LnD57gHaXGgq7O:-LD2Q5utAFzXC_sAHSCd:b-nu8e24) ([raw file](https://github.com/cockroachdb/cockroach/blob/1fe06d67c7e5ad91bc471e1f4ecf847af2b90d5f/pkg/sql/pgwire/pgwire_test.go#L704)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

> when a client doesn't specify a current db in their conn string
> (i.e. in the pgwire options), the defaultdb database is picked up
> instead.

Not sure if it's easy to test.
</blockquote></details>

That is tested by the new tcl test. (`cli/interactive_tests`)

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24735)*

**Comment by BramGruneir** [comment]
Reviewed 23 of 23 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24735)*

**Comment by knz** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/sql/vars.go, line 142 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/24735#-LD2JZrK8tIgvyNG-2em:-LD6xeKcDnwZ-Ej5INV0:b-afh1q2) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4723a9b3db45c7513a0c87f60f6222317470327/pkg/sql/vars.go#L142)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Is there a test for this?
</blockquote></details>

Done. (logic test)

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24735)*

**Pull request merged**